cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -70,6 +73,208 @@ public LineageOperationsProcessor(Set<Connection> stageConnections,
+      // "unusedOperationOutputs.removeAll(stageOutputFields.get(stageName))". Reason for this is
+      // consider the case where OP1 outputs field b, and OP2 outputs the same field b. Operation OP1 is
+      // redundant, since b will be overwritten by OP2. However since field b is  not consumed by any other
+      // operation as an input, redundantOutputs will not have its entry. We still want to fail the validation
+      // for operation OP1.
+      Iterator<Map.Entry<String, List<String>>> iterator = unusedOperationOutputs.entrySet().iterator();
+      while (iterator.hasNext()) {
+        Map.Entry<String, List<String>> next = iterator.next();
+        String field = next.getKey();
+        List<String> origins = next.getValue();
+
+        if (stageOutputFields.get(stageName).contains(field)) {
+          iterator.remove();
+          if (origins.size() > 1) {
+            unusedOperationOutputs.put(field, origins.subList(0, origins.size() - 1));
+          }
+        }
+      }
+
+      unusedOperationOutputs.putAll(redundantOutputs);
+
+      if (!unusedOperationOutputs.isEmpty()) {
+        unusedOutputsFromStage.put(stageName, unusedOperationOutputs);
+      }
+
+      if (!invalidInputs.isEmpty()) {
+        invalidOperationInputs.put(stageName, invalidInputs);
+      }
+    }
+
+    if (unusedOutputsFromStage.isEmpty() && invalidOperationInputs.isEmpty()) {
+      return;
+    }
+
+    String message = invalidOutputErrorMessage(unusedOutputsFromStage);
+    message += invalidInputErrorMessage(invalidOperationInputs);
+    throw new IllegalArgumentException(message);
+  }
+
+  private void processInputOutputs(String operationName, List<String> inputs, List<String> outputs,
+                                   Set<String> validOperationInputs, Map<String, List<String>> invalidInputs,
+                                   Map<String, List<String>> unusedOperationOutputs,
+                                   Map<String, List<String>> redundantOutputs) {
+    for (String field : inputs) {
+      // check if field is valid input. input is valid if it is in the validOperationInputs set
+      if (!validOperationInputs.contains(field)) {
+        List<String> originsWithInvalidInput = invalidInputs.computeIfAbsent(field, k -> new ArrayList<>());
+        originsWithInvalidInput.add(operationName);
+      }
+
+      // current input field can be removed from unusedOperationOutputs set now
+      // as it is being used by current operation
+      List<String> origins = unusedOperationOutputs.get(field);
+      // origins can be null if the field is coming directly from the input schema
+      if (origins != null) {
+        if (origins.size() > 1) {
+          // field is outputted by multiple operations. all occurrences of outputs are redundant
+          // except the last one
+          List<String> redundantOrigins = redundantOutputs.computeIfAbsent(field, k -> new ArrayList<>());
```

Comment:
does this mean we can get here multiple times? If so, doesn't this mean we'll add multiple copies of redundant operations to the redundantOrigins list? Seems like it should be a Set.

Reply:
Since each operation(origin) name is unique, we wont have duplicate operation names in the redundantOrigins for a given field. However for a given field name we can get here again for example: 

```java
op1: [a] -> [x]
op2: [b] -> [x]
op3: [c] -> [x]
op4: [x] -> [y]
```
Now when we are processing input x of op4 (we iterate over operations and for each operation on the input fields), outputsToValidate map will have `x -> [op1, op2, op3]` since its not used by any operation yet. Here op1 and op2 are redundant since the operations are ordered. So in redundantOutputs map we store `x -> [op1, op2]` and we remove entry of x from outputsToValidate.

Now if additional operations are seen in stage as
```java
op5: [a] -> [x]
op6: [b] -> [x]
op7: [c] -> [x]
op8: [x] -> [y]
``` 
while processing x of op8, we find other redundant operations op5 and op6. So we update the redundant set to be `x -> [op1, op2, op5, op6].